miniapps/parelag/MultilevelHcurlSolver.cpp

```diff
@@ -0,0 +1,739 @@
+                                                append("\" -- Build"));
+      if (!myid)
+      {
+         cout << "Building solver \"" << solver_name << "\"...\n";
+      }
+      unique_ptr<Solver> solver = solver_factory->BuildSolver(A, *solver_state);
+      build_timer.Stop();
+
+      // Run the solver.
+      Timer pre_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                              append(solver_name).
+                                              append("\" -- Pre-solve"));
+      if (!myid)
+      {
+         cout << "Solving system with \"" << solver_name << "\"...\n";
+      }
+
+      // Note that X is on true dofs owned by the process, while x is on local
+      // dofs that are known to the process.
+      Vector X(A->Width()), x(sequence[0]->GetNumberOfDofs(jform));
+      X=0.0;
+
+      {
+         Vector tmp(A->Height());
+         A->Mult(X, tmp);
+         tmp *= -1.0;
+         tmp += B;
+
+         double local_norm = tmp.Norml2() * tmp.Norml2();
+         double global_norm;
+         MPI_Reduce(&local_norm, &global_norm, 1, GetMPIType(local_norm),
+                    MPI_SUM, 0, comm);
+         if (!myid)
+         {
+            cout << "Initial residual l2 norm: " << sqrt(global_norm) << '\n';
+         }
+      }
+      pre_timer.Stop();
+
+      // Perform the solve.
+      Timer solve_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                append(solver_name).
+                                                append("\" -- Solve"));
+      solver->Mult(B, X);
+      solve_timer.Stop();
+
+      Timer post_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                               append(solver_name).
+                                               append("\" -- Post-solve"));
+      {
+         Vector tmp(A->Height());
+         A->Mult(X, tmp);
+         tmp *= -1.0;
+         tmp += B;
+
+         double local_norm = tmp.Norml2() * tmp.Norml2();
+         double global_norm;
+         MPI_Reduce(&local_norm, &global_norm, 1, GetMPIType(local_norm),
+                    MPI_SUM, 0, comm);
+         if (!myid)
+         {
+            cout << "Final residual l2 norm: " << sqrt(global_norm) << '\n';
+         }
+      }
+
+      if (!myid)
+      {
+         cout << "Solver \"" << solver_name << "\" finished.\n";
+      }
+
+      // Visualize the solution.
+      if (visualize)
+      {
+         hcurl_dofTrueDof.Distribute(X, x);
+         MultiVector tmp(x.GetData(), 1, x.Size());
+         sequence[0]->show(jform, tmp);
+      }
+      post_timer.Stop();
+   }
+   solvers_timer.Stop();
+
+   total_timer.Stop();
+   TimeManager::Print();
+
+   if (!myid)
+   {
+      cout << "\nFinished.\n";
+   }
+
+   return EXIT_SUCCESS;
+}
+
+// A vector field, used for setting boundary conditions.
+void bdrfunc(const Vector &p, Vector &F)
+{
+   const int dim = p.Size();
+
+   F(0) = 0.0;
```

Comment:
A few lines of code could be saved by just setting `F=0.0;` here and the RHS function.